lang/testkit/src/test/scala/com/wavesplatform/report/QaseReporter.scala

```diff
@@ -0,0 +1,87 @@
+package com.wavesplatform.report
+
+import com.wavesplatform.report.QaseReporter.{QaseProjects, TestResult}
+import io.qase.api.QaseClient
+import io.qase.api.utils.IntegrationUtils
+import io.qase.client.model.ResultCreate
+import org.scalatest.Reporter
+import org.scalatest.events.*
+import play.api.libs.json.{Format, Json}
+
+import java.io.FileWriter
+import java.util.concurrent.ConcurrentHashMap
+import scala.jdk.CollectionConverters.*
+
+class QaseReporter extends Reporter {
+
+  private val results = initResults()
+
+  override def apply(event: Event): Unit = {
+    event match {
+      case ts: TestSucceeded =>
+        extractCaseIds(ts.testName).foreach { case (projectCode, caseId) =>
+          saveTestCaseResults(ResultCreate.StatusEnum.PASSED, ts.testName, projectCode, caseId, None, None, ts.duration)
+        }
+      case tf: TestFailed =>
+        extractCaseIds(tf.testName).foreach { case (projectCode, caseId) =>
+          saveTestCaseResults(ResultCreate.StatusEnum.FAILED, tf.testName, projectCode, caseId, tf.throwable, Some(tf.message), tf.duration)
+        }
+      case _: RunAborted | _: RunStopped | _: RunCompleted =>
+        saveRunResults()
+      case _ => ()
+    }
+  }
+
+  private def extractCaseIds(testName: String): Seq[(String, Long)] = {
+    val patternStr = s"""(${QaseProjects.mkString("|")})-([0-9]+)"""
+    val pattern    = patternStr.r
+
+    pattern.findAllMatchIn(testName).map(m => m.group(1) -> m.group(2).toLong).toSeq
+  }
+
+  private def saveTestCaseResults(
+      status: ResultCreate.StatusEnum,
+      testName: String,
+      projectCode: String,
+      caseId: Long,
+      throwable: Option[Throwable],
+      msgOpt: Option[String],
+      duration: Option[Long]
+  ): Unit =
+    if (QaseClient.isEnabled) {
+      val errMsg     = msgOpt.map(msg => s"\n\n**Error**\n$msg").getOrElse("")
+      val comment    = s"$testName$errMsg"
+      val stacktrace = throwable.map(IntegrationUtils.getStacktrace).orNull
+      val timeMs     = duration.getOrElse(0L)
+
+      results.computeIfPresent(projectCode, (_, results) => TestResult(status.toString, comment, Some(stacktrace), caseId, timeMs) +: results)
```

Comment:
```
val stacktrace = throwable.map(IntegrationUtils.getStacktrace)
```
and usage of `stacktrace` without `Some` is simpler, but probably it made intentionally. If this is true, could you add a comment?

Reply:
Fixed